### DISCUSSION TITLE
Using Lua with C++ (2022-02-14)

### DISCUSSION THREAD

**Discussion by eliasdaler** [eliasdaler/edw.is#3] [Article announcements / comments]
Finally, after 6 years in the making, I give you this:

https://edw.is/using-lua-with-cpp/

Hope it's helpful! Feel free to discuss it here.

**Comment by Nightmare82** [comment]
Very nice article, thanks for that!

I have one question regarding sol : you said one state is usually enough, does that mean you don't support local, per instance variables ? Or is there a way to support that using a single sol::state ? (I currently use one state per script instance)

**Reply by eliasdaler** [reply]
Thank you.

See section named "Try to avoid global variables". If you have local variables, they stay local to the script's scope - other scripts can't access them. You don't need multiple Lua states for that.

If you need to pass local data to Lua from C++, you can do it via function calls, e.g.

```lua
-- in Lua
function someFunc(a, b)
    ...
end
```

```cpp
// in C++
sol::function f = lua["someFunc"];
f(x, y); // no globals are created!
```

If you want to pass variables to *scripts*, it's a bit more complicated. You need to do this:

```cpp
// in C++
sol::load_result script = lua.load_file("script.lua");
script(x, y);
```

```lua
-- script.lua
local x, y = ...
```

Also see [this](https://sol2.readthedocs.io/en/latest/tutorial/all-the-things.html#passing-arguments-to-scripts)

**Reply by eliasdaler** [reply]
As I said in the article, you can treat script files as "functions":

```lua
-- script.lua
local x, y = 5
...
return z
```

is almost equivalent to:
```lua
function script()
    local x, y = 5
    ...
    return z
end
``` 

Read [packages chapter](https://www.lua.org/pil/15.html) to understand it better. :)

**Reply by Nightmare82** [reply]
Ah, nice, thanks for the explanation! Then I can rework it to use a single state indeed.
